x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -888,4 +893,127 @@ public static Expression castStringLiteral(Expression from, DataType target) {
+                mtf.getTypesToIndices().keySet().forEach(typeName -> {
+                    DataType type = DataTypes.fromTypeName(typeName);
+                    TypeResolutionKey key = new TypeResolutionKey(fa.name(), type);
+                    var concreteConvert = typeSpecificConvert(convert, fa.source(), type, mtf);
+                    typeResolutions.put(key, concreteConvert);
+                });
+                // Replace the entire convert function with a new FieldAttribute (containing type conversion knowledge)
+                var resolvedField = resolvedMultiTypeEsField(mtf, typeResolutions);
+                var unionFieldAttribute = new FieldAttribute(fa.source(), fa.name(), resolvedField);  // Generates new ID for the field
+                unionFieldAttributes.add(unionFieldAttribute);
+                return unionFieldAttribute;
+            } else if (convert.field() instanceof AbstractConvertFunction subConvert) {
+                return convert.replaceChildren(Collections.singletonList(resolveConvertFunction(subConvert, unionFieldAttributes)));
+            }
+            return convert;
+        }
+
+        private MultiTypeEsField resolvedMultiTypeEsField(
+            MultiTypeEsField.UnresolvedField mtf,
+            HashMap<TypeResolutionKey, Expression> typeResolutions
+        ) {
+            Map<String, Expression> typesToConversionExpressions = new HashMap<>();
+            mtf.getTypesToIndices().forEach((typeName, indexNames) -> {
+                DataType type = DataTypes.fromTypeName(typeName);
+                TypeResolutionKey key = new TypeResolutionKey(mtf.getName(), type);
+                if (typeResolutions.containsKey(key)) {
+                    typesToConversionExpressions.put(typeName, typeResolutions.get(key));
+                }
+            });
+            return mtf.resolve(typesToConversionExpressions);
+        }
+
+        private Expression typeSpecificConvert(
+            AbstractConvertFunction convert,
+            Source source,
+            DataType type,
+            MultiTypeEsField.UnresolvedField mtf
+        ) {
+            EsField field = new EsField(mtf.getName(), type, mtf.getProperties(), mtf.isAggregatable());
+            NameId id = ((FieldAttribute) convert.field()).id();
+            FieldAttribute resolvedAttr = new FieldAttribute(source, null, field.getName(), field, null, Nullability.TRUE, id, false);
+            return convert.replaceChildren(Collections.singletonList(resolvedAttr));
+        }
+    }
+
+    /**
+     * If there was no AbstractConvertFunction that resolved multi-type fields in the ResolveUnionTypes rules,
+     * then there could still be some FieldAttributes that contain unresolved MultiTypeEsFields.
+     * These need to be converted back to actual UnresolvedAttribute in order for validation to generate appropriate failures.
+     */
+    private static class UnresolveUnionTypes extends AnalyzerRules.AnalyzerRule<LogicalPlan> {
```

Comment:
This rule does something that used to be done by the ResolveRefs code. We have a catch-22 situation where ResolveUnionTypes depends on ResolveRefs to run first, but also relies on InvalidMappedField (or currently MultiTypeEsField.UnresolvedField) to not be converted into an UnresolvedAttribute (which ResolveRefs also does). So that conversion needed to be removed from ResolveRefs and moved after ResolveUnionTypes. This is that new location. This code also did something much more complex before, also editing the contents of EsRelation and the contained EsIndex, but that was getting very messy, and only needed to handle regression tests around unsupported types. I had three choices:

* Continue to make UnresolveUnionTypes more complex (I got all but on regression test passing, so just needed to add nested resolution of Object fields)
* Find a way to get MultiTypeEsField.UnresolvedField to pass through Plan serialization (this is the route I took, but it looks hacky, and I see your comment on that already). This allowed me to reduce this class a lot (you see the simplified version here).
* Merge MultiTypeEsField.UnresolvedField with InvalidMappedField so we avoid the plan serialization hack. This is my new preferred approach, and I see you suggest this too in another comment.

But in no cases does this class completely disappear. Somewhere we need to recognise InvalidMappedField as an unresolved type. That used to happen in ResolveRefs, and now happens here. You are suggesting an alternative location? Verifying EsqlProject? I can investigate that on Monday. I've looked through the Verifier a few times, so I can imagine possibilities there.

Reply:
Thanks - it's a minor point but you could incorporate the rule as part of the UnionType rules in a second pass.
That is convert all InvalidMappedFields that were are not wrapped in a conversion function to the unresolved attribute.
Again minor.